core/drivers/stm32_hpdma.c

```diff
@@ -0,0 +1,349 @@
+			res =
+			stm32_rif_release_semaphore(hpdma_d->base +
+						    _HPDMA_SEMCR(i),
+						    HPDMA_NB_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Couldn't release semaphore for res%u", i);
+				return TEE_ERROR_ACCESS_DENIED;
+			}
+		} else {
+			res =
+			stm32_rif_acquire_semaphore(hpdma_d->base +
+						    _HPDMA_SEMCR(i),
+						    HPDMA_NB_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Couldn't acquire semaphore for res%u", i);
+				return TEE_ERROR_ACCESS_DENIED;
+			}
+		}
+	}
+
+	/*
+	 * Lock RIF configuration if configured. This cannot be undone until
+	 * next reset.
+	 */
+	io_clrsetbits32(hpdma_d->base + _HPDMA_RCFGLOCKR, _HPDMA_RCFGLOCKR_MASK,
+			hpdma_d->conf_data.lock_conf[0]);
+
+end:
+	if (IS_ENABLED(CFG_TEE_CORE_DEBUG)) {
+		/* Check that RIF config are applied, panic otherwise */
+		if ((io_read32(hpdma_d->base + _HPDMA_PRIVCFGR) &
+		     hpdma_d->conf_data.access_mask[0]) !=
+		    hpdma_d->conf_data.priv_conf[0])
+			panic("HPDMA channel priv conf is incorrect");
+
+		if ((io_read32(hpdma_d->base + _HPDMA_SECCFGR) &
+		     hpdma_d->conf_data.access_mask[0]) !=
+		    hpdma_d->conf_data.sec_conf[0])
+			panic("HPDMA channel sec conf is incorrect");
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result parse_dt(const void *fdt, int node,
+			   struct hpdma_pdata *hpdma_d)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int i = 0;
+	int lenp = 0;
+	const fdt32_t *cuint = NULL;
+	struct dt_node_info info = {};
+	struct io_pa_va addr = {};
+
+	fdt_fill_device_info(fdt, &info, node);
+	assert(info.reg != DT_INFO_INVALID_REG &&
+	       info.reg_size != DT_INFO_INVALID_REG_SIZE);
+
+	addr.pa = info.reg;
+	hpdma_d->base = io_pa_or_va_secure(&addr, info.reg_size);
+
+	/* Gate the IP */
+	res = clk_dt_get_by_index(fdt, node, 0, &hpdma_d->hpdma_clock);
+	if (res)
+		return res;
+
+	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
+	if (!cuint)
+		panic("No RIF configuration available");
+
+	hpdma_d->nb_channels = (unsigned int)(lenp / sizeof(uint32_t));
+	assert(hpdma_d->nb_channels <= HPDMA_RIF_CHANNELS);
+
+	hpdma_d->conf_data.cid_confs = calloc(HPDMA_RIF_CHANNELS,
+					      sizeof(uint32_t));
+	hpdma_d->conf_data.sec_conf = calloc(1, sizeof(uint32_t));
+	hpdma_d->conf_data.priv_conf = calloc(1, sizeof(uint32_t));
+	hpdma_d->conf_data.access_mask = calloc(1, sizeof(uint32_t));
+	hpdma_d->conf_data.lock_conf = calloc(1, sizeof(uint32_t));
+	if (!hpdma_d->conf_data.cid_confs || !hpdma_d->conf_data.sec_conf ||
+	    !hpdma_d->conf_data.priv_conf || !hpdma_d->conf_data.access_mask ||
+	    !hpdma_d->conf_data.lock_conf)
+		panic("Missing memory capacity for HPDMA RIF configuration");
+
+	for (i = 0; i < hpdma_d->nb_channels; i++)
+		stm32_rif_parse_cfg(fdt32_to_cpu(cuint[i]), &hpdma_d->conf_data,
+				    HPDMA_NB_MAX_CID_SUPPORTED,
+				    HPDMA_RIF_CHANNELS);
+
+	return TEE_SUCCESS;
+}
+
+static void stm32_hpdma_pm_resume(struct hpdma_pdata *hpdma)
+{
+	apply_rif_config(hpdma, true);
```

Comment:
panic on failure (e.g. failing to acquire a semaphore): there is alternate way on a PM resume operation.

Reply:
I guess you meant "no alternate"

Reply:
Right, sorry, "no alternate way".